File: app/src/main/java/eu/sesma/dagger/di/MainActivityComponent.java
package eu.sesma.dagger.di;

import eu.sesma.dagger.platform.ActivityModule;
import eu.sesma.dagger.platform.MainActivityModule;
import eu.sesma.dagger.ui.main.MainActivity;

import dagger.Subcomponent;
import dagger.android.AndroidInjector;

@PerActivity
@Subcomponent(modules = {ActivityModule.class, MainActivityModule.class})
public interface MainActivityComponent extends AndroidInjector<MainActivity> {

    @Subcomponent.Builder
    abstract class Builder extends AndroidInjector.Builder<MainActivity> {
    }
}

File: module_main/src/main/java/eu/sesma/dagger/main/MainActivityComponent.kt
package eu.sesma.dagger.main

import android.support.v7.app.AppCompatActivity
import dagger.Component
import eu.sesma.dagger.core.ActivityScoped
import eu.sesma.dagger.core.CoreComponent


@ActivityScoped
@Component(
        dependencies = [CoreComponent::class],
        modules = [MainActivityModule::class]
)
interface MainActivityComponent {

    fun inject(activity: MainActivity)

    //Exposed to sub-graphs.
    fun activity(): AppCompatActivity
}